internal/table/client.go

```diff
@@ -295,13 +295,33 @@ func (c *Client) BulkUpsert(
 		return xerrors.WithStackTrace(errClosedClient)
 	}
 
+	request, err := data.ToYDB(tableName)
+	if err != nil {
+		return xerrors.WithStackTrace(err)
+	}
+
+	chunks := make([]*Ydb_Table.BulkUpsertRequest, 0, 1)
+
+	// We must send requests in chunks to avoid exceeding the maximum message size
+	if err := chunkBulkUpsertRequest(&chunks, request, c.config.MaxRequestMessageSize()); err != nil {
+		return xerrors.WithStackTrace(err)
+	}
+
+	return c.sendBulkUpsertRequest(ctx, chunks, opts...)
+}
+
+func (c *Client) sendBulkUpsertRequest(
+	ctx context.Context,
+	requests []*Ydb_Table.BulkUpsertRequest,
+	opts ...table.Option,
+) (finalErr error) {
 	attempts, config := 0, c.retryOptions(opts...)
 	config.RetryOptions = append(config.RetryOptions,
 		retry.WithIdempotent(true),
 		retry.WithTrace(&trace.Retry{
 			OnRetry: func(info trace.RetryLoopStartInfo) func(trace.RetryLoopDoneInfo) {
 				return func(info trace.RetryLoopDoneInfo) {
-					attempts = info.Attempts
+					attempts += info.Attempts - 1
```

Comment:
What about chek if info.Attempts > 0?
If it will be zero - now we will decrement attempts